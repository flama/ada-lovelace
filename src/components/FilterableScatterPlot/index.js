import React, { Component } from 'react'
import PropTypes from 'prop-types'

import ScatterPlot from '../ScatterPlot'
import Categories from '../Categories'
import SelectInput from '../SelectInput'

import './style.scss'

const ALL = "-1"
const radius = 5

class FilterableScatterPlot extends Component {

  constructor(props) {
    super(props)
    this.state = {
      activeRow: ALL,
      activeContinent: ALL
    }
  }

  handleCategoryChange = newCategory => {
    this.setState({
      ...this.state,
      activeRow: this.props.options.categories[newCategory] || ALL
    })
  }

  handleUserSelect = newContinent => {
    this.setState({
      ...this.state,
      activeContinent: newContinent
    })
  }

  filterContinent = data => {
    if(this.state.activeContinent === ALL) return data

    let finalData = {}

    Object.keys(data).forEach(dataKey => {
      let datum = data[dataKey]

      finalData[dataKey] = {
        active: datum.active,
        array: datum.array.filter(woman => woman.Continent === this.state.activeContinent),
        division: {...datum.division}
      }

      Object.keys(datum.division).forEach(divisionKey => {
        let subdivision = datum.division[divisionKey] || []

        finalData[dataKey].division[divisionKey] = subdivision.filter(woman => woman.Continent === this.state.activeContinent)
      })
    })

    return {...data, ...finalData}
  }

  selectCategory = data => {
    if(this.state.activeRow === ALL) return data

    let finalData = {}

    Object.keys(data).forEach(dataKey => {
      finalData[dataKey] = { ...data[dataKey] }

      if(dataKey === this.state.activeRow)
        finalData[dataKey].active = true
    })

    return {...data, ...finalData}
  }

  closeBubble = ev => {
    let contain = (el, list) => {
      if(!el) return true
      for(let i=0; i<list.length; ++i) {
        if(el === list.item(i)) return true
      }
      return false
    }

    ev = ev || {}
    let bubble = document.getElementById('details-bubble')
    let balls = document.getElementsByClassName('d3-point')

    // checks if ball is being clicked
    if(!contain(ev.target, balls))
    {
      for(let i=0; i<balls.length; ++i)
      {
        balls[i].classList.remove('shrinking')
        balls[i].classList.remove('faded')
        balls[i].setAttribute('r', radius)
      }
    }

    bubble && bubble.classList.remove('show')
  }

  render() {
    return (
      <div className="filterable">
        <div className="filters">
          <Categories
            title="Categorias" titlePosition="top" options={this.props.options.categories}
            onChange={ index => this.handleCategoryChange(index) }
            closeBubble={ this.closeBubble }
          />
          <SelectInput options={this.props.options.continents}
            onChange={ index => this.handleUserSelect(index) }
            closeBubble={ this.closeBubble }
          />
        </div>
        <ScatterPlot
          dataList={ this.selectCategory(this.filterContinent(this.props.dataList)) }
          all={ this.state.activeRow === ALL }
          closeBubble={ this.closeBubble }
        />
      </div>
    )
  }
}

FilterableScatterPlot.defaultProps = {
  dataList: {},
  options: {
    categories: []
  }
}

FilterableScatterPlot.propTypes = {
  dataList: PropTypes.object,
  options: PropTypes.shape({
    categories: PropTypes.arrayOf(PropTypes.string).isRequired
  }).isRequired
}

export default FilterableScatterPlot
