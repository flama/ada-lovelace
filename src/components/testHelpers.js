export function fakeElement() {
  let ret = {
    classList: {
      contains: jest.fn(),
      add: jest.fn(),
      remove: jest.fn()
    },
    setAttribute: jest.fn(),
    appendChild: jest.fn(),
    lastChild: true,
    childNodes: [1,2,3],
    removeChild: function() {
      this.lastChild = false
    },
    getBoundingClientRect: () => ({
      top: 500,
      left: 100
    }),
    getElementsByClassName: jest.fn()
      .mockReturnValue({
        item: () => ret
      }),
    style: {},
    id: 'potato-teste/teste'
  }
  return ret
}

export function fakeDocumentGetters(cb) {
  let findByClassName, findById
  let fakeClassNameObject, fakeIdObject, fakeElementsList

  fakeClassNameObject = fakeElement()
  fakeIdObject = fakeElement()

  findByClassName = document.getElementsByClassName
  findById = document.getElementById

  fakeElementsList = [fakeClassNameObject]
  fakeElementsList.item = i => fakeElementsList[i]

  document.getElementsByClassName = jest.fn()
    .mockReturnValue(fakeElementsList)
  document.getElementById = jest.fn()
    .mockReturnValue(fakeIdObject)

  cb({ fakeClassNameObject, fakeIdObject, fakeElementsList })

  afterAll(() => {
    document.getElementsByClassName = findByClassName
    document.getElementById = findById
  })
}
