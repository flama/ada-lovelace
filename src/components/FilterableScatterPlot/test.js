import React from 'react'
import ReactDOM from 'react-dom'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'

import { fakeDocumentGetters } from '../testHelpers'
import FilterableScatterPlot from '../FilterableScatterPlot'

describe("FilterableScatterPlot", () => {
  fakeDocumentGetters(() => {
    let filterable

    let createElementArray = size => {
      const continents = ["potato", "size", "salad"]
      // create an array of `size` length
      return Array.apply(null, Array(size)).map((_, i) => {
        return {
          Continent: continents[i % 3],
          Informations: "x".repeat(size*i)
        }
      })
    }

    beforeEach(() => {
      filterable = mount(
        <FilterableScatterPlot
          dataList={{
            category1: {
              active: false,
              array: createElementArray(5),
              division: {
                subdiv1: createElementArray(3),
                subdiv2: createElementArray(2)
              }
            }
          }}
          options={{
            categories: ["category1"]
          }}
        />
      )
    })

    it("should render correctly", () => {
      expect(toJSON(filterable)).toMatchSnapshot()
    })

    it("properly refreshes on change of continent", () => {
      filterable.find('select').simulate('change')
      filterable.setState({ activeContinent: "salad", ...filterable.state })
      expect(document.getElementById.mock.calls).toEqual([['details-bubble']])
      expect(toJSON(filterable)).toMatchSnapshot()
    })

    it("properly refreshes on change of category", () => {
      filterable.find('button#category1').simulate('click')
      expect(filterable.state("activeRow")).toEqual("category1")
      expect(toJSON(filterable)).toMatchSnapshot()
    })
  })
})
