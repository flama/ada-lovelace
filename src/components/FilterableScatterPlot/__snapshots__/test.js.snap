// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterableScatterPlot properly refreshes on change of category 1`] = `
<FilterableScatterPlot
  dataList={
    Object {
      "category1": Object {
        "active": false,
        "array": Array [
          Object {
            "Continent": "potato",
          },
          Object {
            "Continent": "size",
          },
          Object {
            "Continent": "salad",
          },
          Object {
            "Continent": "potato",
          },
          Object {
            "Continent": "size",
          },
        ],
        "division": Object {
          "subdiv1": Array [
            Object {
              "Continent": "potato",
            },
            Object {
              "Continent": "size",
            },
            Object {
              "Continent": "salad",
            },
          ],
          "subdiv2": Array [
            Object {
              "Continent": "potato",
            },
            Object {
              "Continent": "size",
            },
          ],
        },
      },
    }
  }
  options={
    Object {
      "categories": Array [
        "category1",
      ],
    }
  }
>
  <div
    className="filterable"
  >
    <div
      className="filters"
    >
      <Categories
        all="All"
        closeBubble={[Function]}
        onChange={[Function]}
        options={
          Array [
            "category1",
          ]
        }
        title="Categorias"
        titlePosition="top"
      >
        <div
          className="categories"
        >
          <button
            className="button"
            id="-1"
            onClick={[Function]}
          >
            All
          </button>
          <button
            className="button selected"
            id="category1"
            onClick={[Function]}
          >
            category1
          </button>
        </div>
      </Categories>
      <SelectInput
        closeBubble={[Function]}
        onChange={[Function]}
        options={Array []}
      >
        <div
          className="select-input"
        >
          <select
            onChange={[Function]}
            required="required"
          >
            <option
              value="-1"
            >
              Continents (All)
            </option>
          </select>
        </div>
      </SelectInput>
    </div>
    <ScatterPlot
      all={false}
      closeBubble={[Function]}
      dataList={
        Object {
          "category1": Object {
            "active": true,
            "array": Array [
              Object {
                "Continent": "potato",
              },
              Object {
                "Continent": "size",
              },
              Object {
                "Continent": "salad",
              },
              Object {
                "Continent": "potato",
              },
              Object {
                "Continent": "size",
              },
            ],
            "division": Object {
              "subdiv1": Array [
                Object {
                  "Continent": "potato",
                },
                Object {
                  "Continent": "size",
                },
                Object {
                  "Continent": "salad",
                },
              ],
              "subdiv2": Array [
                Object {
                  "Continent": "potato",
                },
                Object {
                  "Continent": "size",
                },
              ],
            },
          },
        }
      }
    >
      <div
        className="scatter-plot"
        onClick={[Function]}
      >
        <div
          className="topics"
        >
          <ExpansibleScatterRow
            all={false}
            data={
              Object {
                "active": true,
                "array": Array [
                  Object {
                    "Continent": "potato",
                  },
                  Object {
                    "Continent": "size",
                  },
                  Object {
                    "Continent": "salad",
                  },
                  Object {
                    "Continent": "potato",
                  },
                  Object {
                    "Continent": "size",
                  },
                ],
                "division": Object {
                  "subdiv1": Array [
                    Object {
                      "Continent": "potato",
                    },
                    Object {
                      "Continent": "size",
                    },
                    Object {
                      "Continent": "salad",
                    },
                  ],
                  "subdiv2": Array [
                    Object {
                      "Continent": "potato",
                    },
                    Object {
                      "Continent": "size",
                    },
                  ],
                },
              }
            }
            domain={
              Object {
                "x": Array [
                  1800,
                  2017,
                ],
              }
            }
            title="category1"
          >
            <div
              className="expansible-scatter-row"
            >
               
              <ScatterRow
                data={
                  Array [
                    Object {
                      "Continent": "potato",
                    },
                    Object {
                      "Continent": "size",
                    },
                    Object {
                      "Continent": "salad",
                    },
                  ]
                }
                domain={
                  Object {
                    "x": Array [
                      1800,
                      2017,
                    ],
                  }
                }
                hidden={true}
                title="subdiv1"
              >
                <div
                  className="scatter-row hidden"
                >
                  <div
                    className="title"
                  >
                    <h2>
                      subdiv1
                    </h2>
                  </div>
                  <div
                    className="graph"
                  />
                </div>
              </ScatterRow>
              <ScatterRow
                data={
                  Array [
                    Object {
                      "Continent": "potato",
                    },
                    Object {
                      "Continent": "size",
                    },
                  ]
                }
                domain={
                  Object {
                    "x": Array [
                      1800,
                      2017,
                    ],
                  }
                }
                hidden={true}
                title="subdiv2"
              >
                <div
                  className="scatter-row hidden"
                >
                  <div
                    className="title"
                  >
                    <h2>
                      subdiv2
                    </h2>
                  </div>
                  <div
                    className="graph"
                  />
                </div>
              </ScatterRow>
               
            </div>
          </ExpansibleScatterRow>
        </div>
        <DetailsBubble
          close={[Function]}
        >
          <div
            id="details-bubble"
          >
            <a
              className="close"
              onClick={[Function]}
            >
               Close 
            </a>
            <div
              className="content"
            >
              <h3
                className="name"
              />
              <p
                className="description"
              />
            </div>
            <a
              className="external"
              target="_blank"
            >
               More info 
            </a>
          </div>
        </DetailsBubble>
        <div
          className="scatter-row"
        >
          <div
            className="title"
          />
          <GraphTimeline
            className="graph"
            domain={
              Object {
                "x": Array [
                  1800,
                  2017,
                ],
              }
            }
          >
            <div
              className="timeline"
            />
          </GraphTimeline>
        </div>
      </div>
    </ScatterPlot>
  </div>
</FilterableScatterPlot>
`;

exports[`FilterableScatterPlot properly refreshes on change of continent 1`] = `
<FilterableScatterPlot
  dataList={
    Object {
      "category1": Object {
        "active": false,
        "array": Array [
          Object {
            "Continent": "potato",
          },
          Object {
            "Continent": "size",
          },
          Object {
            "Continent": "salad",
          },
          Object {
            "Continent": "potato",
          },
          Object {
            "Continent": "size",
          },
        ],
        "division": Object {
          "subdiv1": Array [
            Object {
              "Continent": "potato",
            },
            Object {
              "Continent": "size",
            },
            Object {
              "Continent": "salad",
            },
          ],
          "subdiv2": Array [
            Object {
              "Continent": "potato",
            },
            Object {
              "Continent": "size",
            },
          ],
        },
      },
    }
  }
  options={
    Object {
      "categories": Array [
        "category1",
      ],
    }
  }
>
  <div
    className="filterable"
  >
    <div
      className="filters"
    >
      <Categories
        all="All"
        closeBubble={[Function]}
        onChange={[Function]}
        options={
          Array [
            "category1",
          ]
        }
        title="Categorias"
        titlePosition="top"
      >
        <div
          className="categories"
        >
          <button
            className="button selected"
            id="-1"
            onClick={[Function]}
          >
            All
          </button>
          <button
            className="button"
            id="category1"
            onClick={[Function]}
          >
            category1
          </button>
        </div>
      </Categories>
      <SelectInput
        closeBubble={[Function]}
        onChange={[Function]}
        options={Array []}
      >
        <div
          className="select-input"
        >
          <select
            onChange={[Function]}
            required="required"
          >
            <option
              value="-1"
            >
              Continents (All)
            </option>
          </select>
        </div>
      </SelectInput>
    </div>
    <ScatterPlot
      all={true}
      closeBubble={[Function]}
      dataList={
        Object {
          "category1": Object {
            "active": false,
            "array": Array [
              Object {
                "Continent": "salad",
              },
            ],
            "division": Object {
              "subdiv1": Array [
                Object {
                  "Continent": "salad",
                },
              ],
              "subdiv2": Array [],
            },
          },
        }
      }
    >
      <div
        className="scatter-plot"
        onClick={[Function]}
      >
        <div
          className="topics"
        >
          <ExpansibleScatterRow
            all={true}
            data={
              Object {
                "active": false,
                "array": Array [
                  Object {
                    "Continent": "salad",
                  },
                ],
                "division": Object {
                  "subdiv1": Array [
                    Object {
                      "Continent": "salad",
                    },
                  ],
                  "subdiv2": Array [],
                },
              }
            }
            domain={
              Object {
                "x": Array [
                  1800,
                  2017,
                ],
              }
            }
            title="category1"
          >
            <div
              className="expansible-scatter-row"
            >
               
              <ScatterRow
                data={
                  Array [
                    Object {
                      "Continent": "salad",
                    },
                  ]
                }
                domain={
                  Object {
                    "x": Array [
                      1800,
                      2017,
                    ],
                  }
                }
                hidden={false}
                title="category1"
              >
                <div
                  className="scatter-row "
                >
                  <div
                    className="title"
                  >
                    <h2>
                      category1
                    </h2>
                  </div>
                  <div
                    className="graph"
                  />
                </div>
              </ScatterRow>
               
            </div>
          </ExpansibleScatterRow>
        </div>
        <DetailsBubble
          close={[Function]}
        >
          <div
            id="details-bubble"
          >
            <a
              className="close"
              onClick={[Function]}
            >
               Close 
            </a>
            <div
              className="content"
            >
              <h3
                className="name"
              />
              <p
                className="description"
              />
            </div>
            <a
              className="external"
              target="_blank"
            >
               More info 
            </a>
          </div>
        </DetailsBubble>
        <div
          className="scatter-row"
        >
          <div
            className="title"
          />
          <GraphTimeline
            className="graph"
            domain={
              Object {
                "x": Array [
                  1800,
                  2017,
                ],
              }
            }
          >
            <div
              className="timeline"
            />
          </GraphTimeline>
        </div>
      </div>
    </ScatterPlot>
  </div>
</FilterableScatterPlot>
`;

exports[`FilterableScatterPlot should render correctly 1`] = `
<FilterableScatterPlot
  dataList={
    Object {
      "category1": Object {
        "active": false,
        "array": Array [
          Object {
            "Continent": "potato",
          },
          Object {
            "Continent": "size",
          },
          Object {
            "Continent": "salad",
          },
          Object {
            "Continent": "potato",
          },
          Object {
            "Continent": "size",
          },
        ],
        "division": Object {
          "subdiv1": Array [
            Object {
              "Continent": "potato",
            },
            Object {
              "Continent": "size",
            },
            Object {
              "Continent": "salad",
            },
          ],
          "subdiv2": Array [
            Object {
              "Continent": "potato",
            },
            Object {
              "Continent": "size",
            },
          ],
        },
      },
    }
  }
  options={
    Object {
      "categories": Array [
        "category1",
      ],
    }
  }
>
  <div
    className="filterable"
  >
    <div
      className="filters"
    >
      <Categories
        all="All"
        closeBubble={[Function]}
        onChange={[Function]}
        options={
          Array [
            "category1",
          ]
        }
        title="Categorias"
        titlePosition="top"
      >
        <div
          className="categories"
        >
          <button
            className="button selected"
            id="-1"
            onClick={[Function]}
          >
            All
          </button>
          <button
            className="button"
            id="category1"
            onClick={[Function]}
          >
            category1
          </button>
        </div>
      </Categories>
      <SelectInput
        closeBubble={[Function]}
        onChange={[Function]}
        options={Array []}
      >
        <div
          className="select-input"
        >
          <select
            onChange={[Function]}
            required="required"
          >
            <option
              value="-1"
            >
              Continents (All)
            </option>
          </select>
        </div>
      </SelectInput>
    </div>
    <ScatterPlot
      all={true}
      closeBubble={[Function]}
      dataList={
        Object {
          "category1": Object {
            "active": false,
            "array": Array [
              Object {
                "Continent": "potato",
              },
              Object {
                "Continent": "size",
              },
              Object {
                "Continent": "salad",
              },
              Object {
                "Continent": "potato",
              },
              Object {
                "Continent": "size",
              },
            ],
            "division": Object {
              "subdiv1": Array [
                Object {
                  "Continent": "potato",
                },
                Object {
                  "Continent": "size",
                },
                Object {
                  "Continent": "salad",
                },
              ],
              "subdiv2": Array [
                Object {
                  "Continent": "potato",
                },
                Object {
                  "Continent": "size",
                },
              ],
            },
          },
        }
      }
    >
      <div
        className="scatter-plot"
        onClick={[Function]}
      >
        <div
          className="topics"
        >
          <ExpansibleScatterRow
            all={true}
            data={
              Object {
                "active": false,
                "array": Array [
                  Object {
                    "Continent": "potato",
                  },
                  Object {
                    "Continent": "size",
                  },
                  Object {
                    "Continent": "salad",
                  },
                  Object {
                    "Continent": "potato",
                  },
                  Object {
                    "Continent": "size",
                  },
                ],
                "division": Object {
                  "subdiv1": Array [
                    Object {
                      "Continent": "potato",
                    },
                    Object {
                      "Continent": "size",
                    },
                    Object {
                      "Continent": "salad",
                    },
                  ],
                  "subdiv2": Array [
                    Object {
                      "Continent": "potato",
                    },
                    Object {
                      "Continent": "size",
                    },
                  ],
                },
              }
            }
            domain={
              Object {
                "x": Array [
                  1800,
                  2017,
                ],
              }
            }
            title="category1"
          >
            <div
              className="expansible-scatter-row"
            >
               
              <ScatterRow
                data={
                  Array [
                    Object {
                      "Continent": "potato",
                    },
                    Object {
                      "Continent": "size",
                    },
                    Object {
                      "Continent": "salad",
                    },
                    Object {
                      "Continent": "potato",
                    },
                    Object {
                      "Continent": "size",
                    },
                  ]
                }
                domain={
                  Object {
                    "x": Array [
                      1800,
                      2017,
                    ],
                  }
                }
                hidden={false}
                title="category1"
              >
                <div
                  className="scatter-row "
                >
                  <div
                    className="title"
                  >
                    <h2>
                      category1
                    </h2>
                  </div>
                  <div
                    className="graph"
                  />
                </div>
              </ScatterRow>
               
            </div>
          </ExpansibleScatterRow>
        </div>
        <DetailsBubble
          close={[Function]}
        >
          <div
            id="details-bubble"
          >
            <a
              className="close"
              onClick={[Function]}
            >
               Close 
            </a>
            <div
              className="content"
            >
              <h3
                className="name"
              />
              <p
                className="description"
              />
            </div>
            <a
              className="external"
              target="_blank"
            >
               More info 
            </a>
          </div>
        </DetailsBubble>
        <div
          className="scatter-row"
        >
          <div
            className="title"
          />
          <GraphTimeline
            className="graph"
            domain={
              Object {
                "x": Array [
                  1800,
                  2017,
                ],
              }
            }
          >
            <div
              className="timeline"
            />
          </GraphTimeline>
        </div>
      </div>
    </ScatterPlot>
  </div>
</FilterableScatterPlot>
`;
